pom.xml

```diff
@@ -129,6 +129,11 @@
       <version>5.1.0</version>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>com.google.guava</groupId>
+      <artifactId>guava</artifactId>
+      <version>31.1-android</version>
+    </dependency>
```

Comment:
:thinking: here - in the `<dependencyManagement>` section of this file, can you try removing the parts where we pin the dependencies for the Jenins plugins? (this bit below):

```xml
    <dependency>
      <groupId>org.jenkins-ci.plugins</groupId>
      <artifactId>mailer</artifactId>
      <version>463.vedf8358e006b_</version>
    </dependency>
    <dependency>
      <groupId>org.jenkins-ci.plugins</groupId>
      <artifactId>matrix-auth</artifactId>
      <version>3.2.1</version>
    </dependency>
    <dependency>
      <groupId>org.jenkins-ci.plugins</groupId>
      <artifactId>credentials</artifactId>
      <version>1311.vcf0a_900b_37c2</version>
    </dependency>
```

This will a) ensure that we are using newer versions of the upstream plugins, and b) potentially remove the need to declare `guava` a managed transitive dependency.

Reply:
I have tried that.
```
git --no-pager diff
diff --git a/pom.xml b/pom.xml
index 58427f4..ce36b4d 100644
--- a/pom.xml
+++ b/pom.xml
@@ -108,32 +108,12 @@
       <artifactId>google-oauth-client-java6</artifactId>
       <version>1.35.0</version>
     </dependency>
-    <dependency>
-      <groupId>org.jenkins-ci.plugins</groupId>
-      <artifactId>mailer</artifactId>
-      <version>463.vedf8358e006b_</version>
-    </dependency>
-    <dependency>
-      <groupId>org.jenkins-ci.plugins</groupId>
-      <artifactId>matrix-auth</artifactId>
-      <version>3.2.1</version>
-    </dependency>
-    <dependency>
-      <groupId>org.jenkins-ci.plugins</groupId>
-      <artifactId>credentials</artifactId>
-      <version>1311.vcf0a_900b_37c2</version>
-    </dependency>
     <dependency>
       <groupId>org.easymock</groupId>
       <artifactId>easymock</artifactId>
       <version>5.1.0</version>
       <scope>test</scope>
     </dependency>
-    <dependency>
-      <groupId>com.google.guava</groupId>
-      <artifactId>guava</artifactId>
-      <version>31.1-android</version>
-    </dependency>
     </dependencies>
   </dependencyManagement>
```
However, it throws the below error.
```
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-enforcer-plugin:3.4.1:enforce (display-info) on project openshift-login:
[ERROR] Rule 5: org.apache.maven.enforcer.rules.dependency.RequireUpperBoundDeps failed with message:
[ERROR] Failed while enforcing RequireUpperBoundDeps. The error(s) are [
[ERROR] Require upper bound dependencies error for com.google.errorprone:error_prone_annotations:2.18.0 paths to dependency are:
[ERROR] +-org.openshift.jenkins:openshift-login:1.1.0.999999-SNAPSHOT
[ERROR]   +-com.google.http-client:google-http-client-jackson2:1.43.3
[ERROR]     +-com.google.http-client:google-http-client:1.43.3
[ERROR]       +-com.google.errorprone:error_prone_annotations:2.18.0 (managed) <-- com.google.errorprone:error_prone_annotations:2.18.0
[ERROR] and
[ERROR] +-org.openshift.jenkins:openshift-login:1.1.0.999999-SNAPSHOT
[ERROR]   +-com.google.http-client:google-http-client-jackson2:1.43.3
[ERROR]     +-com.google.http-client:google-http-client:1.43.3
[ERROR]       +-com.google.guava:guava:33.0.0-jre (managed) <-- com.google.guava:guava:30.1.1-android
[ERROR]         +-com.google.errorprone:error_prone_annotations:2.18.0 (managed) <-- com.google.errorprone:error_prone_annotations:2.23.0
[ERROR] ]
[ERROR] -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```
Declaring the managed transitive dependency `com.google.errorprone:error_prone_annotations:2.23.0` resolved the issue.
```
git --no-pager diff
diff --git a/pom.xml b/pom.xml
index 58427f4..6f6523a 100644
--- a/pom.xml
+++ b/pom.xml
@@ -108,21 +108,6 @@
       <artifactId>google-oauth-client-java6</artifactId>
       <version>1.35.0</version>
     </dependency>
-    <dependency>
-      <groupId>org.jenkins-ci.plugins</groupId>
-      <artifactId>mailer</artifactId>
-      <version>463.vedf8358e006b_</version>
-    </dependency>
-    <dependency>
-      <groupId>org.jenkins-ci.plugins</groupId>
-      <artifactId>matrix-auth</artifactId>
-      <version>3.2.1</version>
-    </dependency>
-    <dependency>
-      <groupId>org.jenkins-ci.plugins</groupId>
-      <artifactId>credentials</artifactId>
-      <version>1311.vcf0a_900b_37c2</version>
-    </dependency>
     <dependency>
       <groupId>org.easymock</groupId>
       <artifactId>easymock</artifactId>
@@ -130,9 +115,9 @@
       <scope>test</scope>
     </dependency>
     <dependency>
-      <groupId>com.google.guava</groupId>
-      <artifactId>guava</artifactId>
-      <version>31.1-android</version>
+      <groupId>com.google.errorprone</groupId>
+      <artifactId>error_prone_annotations</artifactId>
+      <version>2.23.0</version>
     </dependency>
     </dependencies>
   </dependencyManagement>
```
Should we check any of the other possible combinations?

Reply:
I have made that change. Another question: how can we differentiate between large and small libraries? Can we use the file size information available on Maven Repository to do this?

Reply:
> Can we use the file size information available on Maven Repository to do this?

Perhaps. This is honestly more "art" than science, and experienced Java developers tend to have an intuitive sense of which dependencies are large/commonly referenced.